result.go

```diff
@@ -72,19 +70,89 @@ func newResult(t reflect.Type) (result, error) {
 	}
 }
 
+// resultVisitor visits every result in a result tree, allowing tracking state
+// at each level.
+type resultVisitor interface {
+	// Visit is called on the result being visited.
+	//
+	// If Visit returns a non-nil resultVisitor, that resultVisitor visits all
+	// the child results of this result.
+	Visit(result) resultVisitor
+
+	// VisitField is called on each field of a resultObject after visiting it
+	// but before walking its descendants.
+	//
+	// The same resultVisitor is used for all fields: the one returned upon
+	// visiting the resultObject.
+	//
+	// For each visited field, if VisitField returns a non-nil resultVisitor,
+	// it will be used to walk the result of that field.
+	VisitField(resultObjectField) resultVisitor
```

Comment:
I almost want to call these "AnnotateWithField" and "AnnotateWithPosition" because they're not reaaaaaally 'visiting' those fields for any other purpose.  Especially VisitPosition, there's not much you could do with it other than annotate (since it's just an index).  If we did something with it we'd have to keep state in the visitor object itself which could be the grounds for lots of confusion...

Reply:
Good call. Renaming.